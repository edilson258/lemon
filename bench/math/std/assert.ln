
pub fn assert(value: bool) = {
  if (!value) panic("assertion failed");
}
