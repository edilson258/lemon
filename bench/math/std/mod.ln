pub const max = import("max.ln");

// const max = fn(a: i32, b: i32): i32  = {
//   if (a > b) return a;
//   return b;
// };

// pub fn check_max(left: i32, right: i32, expected: i32) = {
//   let test = max(left, right);
// }
