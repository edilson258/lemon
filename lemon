#!/bin/bash

./target/release/lemon "$@"
