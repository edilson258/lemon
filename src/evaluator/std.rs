use std::collections::HashMap;

use crate::evaluator::formatting::display_value;

use super::value::{create_null, Value};

pub type StdFn = fn(Vec<Value>) -> Value;

pub fn create_std_fn() -> HashMap<&'static str, StdFn> {
  let mut map: HashMap<&'static str, StdFn> = HashMap::new();
  // io
  map.insert("println", |args| {
    let output = args.iter().map(|arg| display_value(arg).to_string()).collect::<Vec<String>>().join("");
    println!("{}", output);
    create_null()
  });
  map.insert("print", |args| {
    let output = args.iter().map(|arg| display_value(arg).to_string()).collect::<Vec<String>>().join(" ");
    println!("{}", output);
    create_null()
  });

  // network
  return map;
}
