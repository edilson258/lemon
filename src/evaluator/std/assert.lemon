
fn assert(x, err) = match x {
  true => null,
  _ => {
    println(err);
    std::exit(1);
  }
};
