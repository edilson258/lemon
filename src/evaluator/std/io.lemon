
let io = import("core::io");
let fmt = import("core::fmt");


fn println(x) = {
  io::write_stdout(fmt::format("{}\n", x));
};

fn print(x) = {
  io::write_stdout(x);
};
