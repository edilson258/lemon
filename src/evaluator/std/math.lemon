
fn double(num) = match num {
  1 => 2,
  _ => num * 2,
};

fn add(rt, lt) = rt + lt;

fn sub(rt, lt) = rt - lt;

fn mul(rt, lt) = rt * lt;

fn div(rt, lt) = rt / lt;

fn rem(rt, lt) = rt % lt;


fn min(rt, lt) = match rt < lt {
  true => rt,
  false => lt,
};


fn max(rt, lt) = match rt > lt {
  true => rt,
  false => lt,
};


fn abs(num) = match num > 0 {
  true => num,
  false => num * -1,
};


fn bit_or(rt, lt) = rt | lt;

fn round(x) = x + 0.5 |> bit_or(0);

fn floor(x) = x |> bit_or(0)
