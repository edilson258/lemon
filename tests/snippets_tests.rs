use std::{
  collections::HashMap,
  fs,
  path::{Path, PathBuf},
};

use insta::assert_snapshot;
use lemon::{diag::Diag, lexer::Token, parser::Parser, range::Range, source::Source};
use logos::Logos;
use stdext::function_name;
use walkdir::WalkDir;

type RunFn = dyn Fn(&str, &Path) -> Result<String, Diag>;

const ROOT_TEST_FOLDER: &str = "/tests/snippets/";

fn run_single_snippet(path: &Path, run: &[&RunFn]) -> Result<(), String> {
  println!("{}", path.display());
  let code = fs::read_to_string(path).map_err(|e| e.to_string())?;
  let file_name = path
    .to_str()
    .and_then(|p| p.rsplit_once(ROOT_TEST_FOLDER))
    .map(|(_, name)| name)
    .ok_or_else(|| "not found".to_string())?;

  let file_path = format!("{}{}", &ROOT_TEST_FOLDER[1..], file_name);
  let file_path = Path::new(file_path.as_str());
  let mut results: HashMap<&Path, Vec<String>> = HashMap::new();

  for fun in run {
    let result = fun(&code, file_path).unwrap_or_else(|err| err.to_string());
    results.entry(file_path).or_default().push(result);
  }

  let results = results.into_values().map(|v| v.join("\n")).collect::<Vec<_>>();

  let mut settings = insta::Settings::clone_current();
  settings.set_prepend_module_to_snapshot(false);
  settings.set_omit_expression(true);
  settings.set_input_file(path);

  settings.bind(|| {
    for result in results {
      assert_snapshot!(file_name, result);
    }
  });

  Ok(())
}

fn run_snippets_dir_multiple(test_name: &str, run: &[&RunFn]) {
  let root = PathBuf::from(format!(
    "{}{ROOT_TEST_FOLDER}{}",
    env!("CARGO_MANIFEST_DIR"),
    test_name.rsplit_once(':').unwrap().1
  ));
  println!("Root: {}", root.display());

  let walker = WalkDir::new(&root).sort_by_file_name().max_depth(1).into_iter().filter_entry(|e| {
    let path = e.path();
    path == root
      || path.is_dir()
      || (path.is_file() && path.extension().map_or(false, |x| x == "lemon" || x == "ln"))
  });

  for entry in walker {
    let entry = entry.unwrap();
    let path = entry.path();
    if path.is_file() {
      eprintln!("Testing {}", path.display());
      run_single_snippet(path, run).unwrap();
    }
  }
}

fn run_snippets_dir(test_name: &str, run: &RunFn) {
  println!("Running snippets dir: {}", test_name);
  run_snippets_dir_multiple(test_name, &[run])
}

fn show_pretty_lexer(result: &[(Range, Token, String)]) -> String {
  let mut pretty = String::new();
  for (i, (range, token, token_text)) in result.iter().enumerate() {
    pretty.push_str(&format!(
      "- entry {}:\n    range: {}\n    token: {:?}\n    lexeme: {}\n",
      i + 1,
      range,
      token,
      token_text
    ));
  }
  pretty
}
#[test]
fn lexer() {
  run_snippets_dir(function_name!(), &|code, path| {
    let source = Source::new(code.to_string(), path.to_path_buf());
    let mut lexer = Token::lexer(source.raw());
    let mut result = Vec::new();

    while let Some(Ok(token)) = lexer.next() {
      let range = Range::from_span(lexer.span());
      let token_text = lexer.slice().to_string();
      result.push((range, token, token_text));
    }
    Ok(show_pretty_lexer(&result))
  });
}

#[test]
fn parser() {
  run_snippets_dir(function_name!(), &|code, path| {
    let source = Source::new(code.to_string(), path.to_path_buf());
    let mut lex = Token::lexer(source.raw());
    let mut parser = Parser::new(&mut lex);
    let result = parser.parse_program();
    if let Err(err) = result {
      return Ok(format!("{:?}", err));
    }
    if let Ok(program) = result {
      return Ok(format!("{:#?}", program));
    }
    Ok(String::new())
  });
}
