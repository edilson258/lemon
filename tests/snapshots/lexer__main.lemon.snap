---
source: tests/snippets_tests.rs
input_file: tests/snippets/lexer/main.lemon
---
- entry 1:
    range: 3..5
    token: Fn
    lexeme: fn
- entry 2:
    range: 6..10
    token: Ident
    lexeme: main
- entry 3:
    range: 10..11
    token: LParen
    lexeme: (
- entry 4:
    range: 11..12
    token: RParen
    lexeme: )
- entry 5:
    range: 13..14
    token: Assign
    lexeme: =
- entry 6:
    range: 15..16
    token: LBrace
    lexeme: {
- entry 7:
    range: 18..19
    token: RBrace
    lexeme: }
